internal/controller/controllers/agent_controller.go

```diff
@@ -72,6 +80,30 @@ func (r *AgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 		r.Log.Debugf("ClusterDeploymentName not set in Agent %s. Skipping Reconcile", agent.Name)
 		return ctrl.Result{Requeue: false}, nil
 	}
+
+	if agent.ObjectMeta.DeletionTimestamp.IsZero() { // agent no being deleted
+		// Register a finalizer if it is absent.
+		if !funk.ContainsString(agent.GetFinalizers(), AgentFinalizerName) {
+			controllerutil.AddFinalizer(agent, AgentFinalizerName)
+			if err = r.Update(ctx, agent); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else { // agent is being deleted
+		if funk.ContainsString(agent.GetFinalizers(), AgentFinalizerName) {
+			// deletion finalizer found, deregister the backend host and delete the agent
+			reply, cleanUpErr := r.deregisterHostIfNeeded(ctx, req.NamespacedName)
+			if cleanUpErr != nil {
+				return reply, cleanUpErr
+			}
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(agent, AgentFinalizerName)
+			if err = r.Update(ctx, agent); err != nil {
```

Comment:
what happens if a host deregistered successfully but for some reason, the update fails? will the finalizer still be removed or it takes effect only after the update?

Reply:
if we fail to update, that means that the object still includes the finalized.
in that case, we return `err` which is not `nil` so  it will requeue and try to remove it (and update) again